parser_library/src/processing/statement_analyzers/hit_count_analyzer.cpp

```diff
@@ -60,33 +66,40 @@ std::optional<range> get_range(const context::hlasm_statement* stmt,
     return std::nullopt;
 }
 
-size_t& emplace_hit_count(
-    hit_count_map& hc_map, const utils::resource::resource_location& rl, range r, bool can_be_external_macro)
+void emplace_hit_count(hit_count_map& hc_map,
+    const utils::resource::resource_location& rl,
+    stmt_lines_range lines_range,
+    bool can_be_external_macro,
+    bool increase_hit_count)
 {
-    static constexpr auto map_emplacer = []<typename VALUE>(auto& m, const auto& k) {
-        if (auto m_it = m.find(k); m_it != m.end())
-            return m_it;
+    auto& start_line = lines_range.first;
+    auto& end_line = lines_range.second;
 
-        return m.try_emplace(k, VALUE {}).first;
-    };
+    auto& stmt_hc_details = hc_map.try_emplace(rl).first->second;
+    stmt_hc_details.is_external_macro &= can_be_external_macro;
+    stmt_hc_details.max_lineno = std::max(stmt_hc_details.max_lineno, start_line);
 
-    auto& stmt_hc_details = map_emplacer.template operator()<stmt_hit_count_details>(hc_map, rl)->second;
-    auto& [stmt_r, stmt_count] =
-        map_emplacer.template operator()<hit_count_details>(stmt_hc_details.stmt_hc_map, r.start.line)->second;
+    auto& line_hits = stmt_hc_details.line_hits;
+    while (line_hits.size() <= end_line)
+        line_hits.resize(2 * line_hits.size());
 
-    stmt_hc_details.is_external_macro &= can_be_external_macro;
-    stmt_r = std::move(r);
-    return stmt_count;
+    std::for_each_n(
+        std::next(line_hits.begin(), start_line), end_line - start_line + 1, [increase_hit_count](auto& line_hit) {
+            line_hit.contains_statement = true;
+
+            if (increase_hit_count)
+                line_hit.count++;
+        });
 }
 } // namespace
 
 void hit_count_analyzer::emplace_macro_header_definitions(const context::id_index& id)
 {
-    if (auto map_it = m_macro_header_definitions.find(id.to_string()); map_it != m_macro_header_definitions.end())
+    if (auto map_it = m_macro_header_definitions.find(id); map_it != m_macro_header_definitions.end())
     {
         auto& mac_header_details = map_it->second;
-        emplace_hit_count(m_hit_counts, mac_header_details.rl, std::move(mac_header_details.init_r), true)++;
-        emplace_hit_count(m_hit_counts, mac_header_details.rl, std::move(mac_header_details.name_r), true)++;
+        emplace_hit_count(m_hit_counts, mac_header_details.rl, std::move(mac_header_details.init_line_r), true, true);
+        emplace_hit_count(m_hit_counts, mac_header_details.rl, std::move(mac_header_details.name_line_r), true, true);
```

Comment:
The move here makes no difference, but in principle I am not sure it should be there anyway.

Also I wonder if the function should just take the ref to m_hit_counts.try_emplace()->first.second, instread of the two parameters (this action could be wrapped in a member func taking the const resource_location&)

Reply:
Created new function and move is no longer there